Make is_equal_approx separate and make == exact again

aaronfranke opened the pull request:
This PR fixes the concerns @reduz has about #18992, and fixes #32604

My initial reasoning is that end-user code would almost always want to have equality checks be safe, so that would make sense to have as the default behavior of `==` for vectors. One concern about this is with performance, but if that was the only concern, performance-critical code could have a separate method for checking this (and some code could also switch to integer vectors later on, too).

This PR restores the old (3.1) functionality of `==` for vectors etc, and instead introduces some `is_equal_approx` methods for comparing them. The main reason I think this is the best course of action is that, as @reduz mentioned, we don't want to hide this functionality from the user. It makes sense to require users to be aware of approximate equality and consciously choose it, instead of it just being the default choice for `==`. Aside from that, it does also have the advantage of allowing code to run as fast as before.

The first commit adds `is_equal_approx` internally (exposed in C#) to the types (so, nothing breaking). There is a pre-existing exposed `is_equal_approx` in `Basis` that should be replaced in 4.0, I've placed TODO notes in there. The second commit switches internal code that depends on approximate vector equality (such as the lines in #31395) to use the `is_equal_approx` methods (so, nothing breaking). The third commit exposes `is_equal_approx` methods and reverts `==` to check for exact equality again (same behavior as 3.1, so 3.1 projects will still work the same).

One thing I wasn't sure about is with `>` and `>=` etc. I think it's desirable to leave these as they currently are, but I could be wrong.

As a side note, this is my **100th** pull request on GitHub (and my 51st to the Godot repo). :tada:

akien-mga commented on core/math/plane.cpp:
```diff
@@ -157,8 +154,13 @@ bool Plane::intersects_segment(const Vector3 &p_begin, const Vector3 &p_end, Vec
 /* misc */
 
 bool Plane::is_almost_like(const Plane &p_plane) const {
+	// TODO: Remove this in Godot 4.0.
```
It's not exposed, so it can be removed already.

aaronfranke replied:
Done.

aaronfranke committed:
Replace vector == and is_zero_approx(distance) with is_equal_approx

Internal changes only

aaronfranke committed:
Expose is_equal_approx and restore == to be exact again

This commit changes behavior for GDScript and C#.

Also did some organizing of the order to logically group related methods, mostly for Rect2 and AABB.

akien-mga commented:
I would really prefer a PR that *only* addresses the issues raised in #18992.

It's unclear whether many of the added approximation checks are necessary, and as mentioned by @reduz, they might impair the compiler's ability to perform optimizations.

AFAICT reverting the changes to operators in `Plane`, `Vector2` and `Vector3` should be enough for now. The rest can be added in a second step, in a PR that would likely not be merged for 3.2 as it's too late.

reduz commented on core/math/basis.cpp:
```diff
@@ -557,28 +557,9 @@ void Basis::set_euler_yxz(const Vector3 &p_euler) {
 	*this = ymat * xmat * zmat;
 }
 
-bool Basis::is_equal_approx(const Basis &a, const Basis &b, real_t p_epsilon) const {
```
The epsilon is needed afaik because it can be used with different values for different use cases, when matrix is not of unit size. Check those that use is_equal_approx_ratio and leave them alone.

aaronfranke replied:
Should I leave a note that it would be worth renaming this method to `is_equal_approx_ratio` in 4.0? It would be strange if Basis had different behavior for `is_equal_approx`.

reduz reviewed (changes requested):
leave == comparisons alone if they did not use is_equal_approx originally.

reduz commented:
ack stupid github, dont merge this, i only approved the last commit.

aaronfranke commented on core/math/delaunay.h:
```diff
@@ -80,11 +80,11 @@ class Delaunay2D {
 	}
 
 	static bool edge_compare(const Vector<Vector2> &p_vertices, const Edge &p_a, const Edge &p_b) {
-		if (p_vertices[p_a.edge[0]] == p_vertices[p_b.edge[0]] && p_vertices[p_a.edge[1]] == p_vertices[p_b.edge[1]]) {
```
They were *not* originally like this. This is a reversion of https://github.com/godotengine/godot/pull/31395/files#diff-1c96941701be95512a1964a0e4e5c5c4L83

aaronfranke commented on core/math/plane.cpp:
```diff
@@ -156,11 +153,6 @@ bool Plane::intersects_segment(const Vector3 &p_begin, const Vector3 &p_end, Vec
 
 /* misc */
 
-bool Plane::is_almost_like(const Plane &p_plane) const {
```
This is only used internally, @akien-mga told me to remove it.

akien-mga replied:
Yeah it's ok to remove this one.

aaronfranke commented on modules/csg/csg.cpp:
```diff
@@ -242,7 +242,7 @@ void CSGBrushOperation::BuildPoly::_clip_segment(const CSGBrush *p_brush, int p_
 	//check if edge and poly share a vertex, of so, assign it to segment_idx
 	for (int i = 0; i < points.size(); i++) {
 		for (int j = 0; j < 2; j++) {
-			if (segment[j] == points[i].point) {
```
They were originally `is_zero_approx(a.distance_to(b))` which is functionally identical to checking for approximate equality except that the old code doesn't work with large vectors. This is necessary to keep #22988 fixed.

aaronfranke committed:
Make is_equal_approx separate for structures

This commit adds exposed behavior for C#

akien-mga commented:
As mentioned, I would *really* have preferred a PR only addressing the necessary bug fixes, i.e. undoing approximate equality checks added where exact checks are meant to be used.

This PR does that, and adds more utility methods, some cleanup, some refactoring... it makes it big and quite inappropriate to merge at such a late stage in development.

Still, the bug fixes are important and I don't feel like doing a slimmed down version of the PR with only fixes myself. After the last revision and discussing it with @reduz, we agree that the extra enhancements should likely be fine, so let's merge.

akien-mga commented:
Thanks!

aaronfranke commented:
@akien-mga For the record, having only the changes to the `==` operator would break other things, and I'm not keen on doing that, or creating workarounds in the affected code, just to minimize changes. I'd rather create a comprehensive solution, so I did.

If any further issues are discovered due to merging such a large PR I'll get on it immediately.
